content/posts/2025-09-01-gsoc-ubsan-trap-messages.md

```diff
@@ -0,0 +1,410 @@
+add.c:5:13: runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'
+SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior add.c:5:13 
+Added 2147483647 + 1 = -2147483648
+```
+
+
+In contrast when using UBSan in trapping mode the program immediately terminates when undefined behavior is detected as shown below.
+
+
+```
+$ clang -fsanitize=undefined -fsanitize-trap=undefined add.c -g -o add && ./add 2147483647 1
+[1]    54357 trace trap  ./add 2147483647 1
+```
+
+
+This is the expected behavior of trapping mode but how should a developer debug what happened when a trap is hit? If we attach a debugger and run the example program this is the output LLDB shows.
+
+
+```
+$ lldb ./add -- 2147483647 1    
+(lldb) target create "./add"
+(lldb) settings set -- target.run-args  "2147483647" "1"
+(lldb) r
+Process 17347 launched: 'add' (arm64)
+Process 17347 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100003d3c)
+    frame #0: 0x0000000100003d3c add`add(a=2147483647, b=1) at add.c:5:13
+   2    #include <stdio.h>
+   3   
+   4    int add(int a, int b) {
+-> 5        return a + b;
+   6    }
+   7   
+   8    int main(int argc, const char** argv) {
+(lldb) bt
+* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100003d3c)
+  * frame #0: 0x0000000100003d3c add`add(a=2147483647, b=1) at add.c:5:13
+    frame #1: 0x0000000100003eec add`main(argc=3, argv=0x000000016fdff110) at add.c:13:18
+    frame #2: 0x00000001842bab98 dyld`start + 6076
+(lldb) dis -p
+add`add:
+->  0x100003d3c <+40>: brk    #0x5500
+    0x100003d40 <+44>: ldr    w0, [sp, #0x4]
+    0x100003d44 <+48>: add    sp, sp, #0x10
+    0x100003d48 <+52>: ret   
+```
+
+
+We can see that a `brk` instruction was hit while handling the `a + b` expression but there is no good explanation of what happened. `brk` is the trap instruction on arm64 but it is not particularly clear this has anything to do with UBSan. For this toy example we can speculate that integer overflow occurred because the program was built with trapping UBSan and the trap was hit while handling `a + b`. However, in real programs built with trapping UBSan and potentially other hardening mechanisms it is often far less obvious what happened.
+
+
+For this particular example. The information that this is an integer overflow UBSan check is actually there but it is not very obvious. On x86_64 and arm64 the reason for trapping is actually [encoded in the operand to the trap instruction](https://maskray.me/blog/2023-01-29-all-about-undefined-behavior-sanitizer ). In this case the `#0x5500` immediate to the brk instruction encodes that this is a UBSan trap for integer overflow. The UBSan immediate is encoded as `('U' << 8) + SanitizerHandler` where `SanitizerHandler` is the enum value from the `SanitizerHandler` enum inside Clang’s internals.
```

Comment:
Should we provide a link to the [`SanitizerHandler` enum](https://github.com/llvm/llvm-project/blob/96195e7d44613e272475c90df187678036f21966/clang/lib/CodeGen/SanitizerHandler.h#L78)?